src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -494,7 +494,7 @@ public ParsedDocument parse(SourceToParse source, @Nullable ParseListener listen
         } catch (Throwable e) {
             // if its already a mapper parsing exception, no need to wrap it...
             if (e instanceof MapperParsingException) {
-                throw (MapperParsingException) e;
+                throw ((MapperParsingException) e).setMappingsModified(context.mappingsModified());
```

Comment:
Could we do it in such a way that we do not need to modify the incoming exception in-place?